### PR TITLE
Detached validator from node process

### LINKED ISSUE #165: Option to detach validator from node process so it doesn't take time restarting during development
Will require also providing the user with a way to manually kill the validator when they're done with their session.

### PR DESCRIPTION
**Pull request by JoshL18**
(closes #165)

HTML validator can now optionally be detached from node process, this change:

- Allows validator to continuously run without taking time to restart by detaching it from the node process.
- Created new HTML validator kill script which can be executed by calling `npm run killValidator`, the kill script will find the validator and kill it no matter what port it is on.
- You can now run the validator in `detached` mode by running the command `npm run dev detach`.
- You can now specify that you want the node process to have an `attached` validator by typing `npm run dev attach`, though the validator is `attached` by default.
- Calling `npm run dev attach` after `npm run dev detach` will kill the `detached` validator and start a new one `attached` to the node process.
- You can use `export ENABLE_DETACH=true && npm run dev` to force the validator to run `detached` from the node process.
- You can use `export ENABLE_DETACH=false && npm run dev` to force the validator to run `attached` to the node process.
- Updated README to account for these changes.

### CODE REVIEW

File: lib/sourceParams.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ module.exports = function(app) {
     disableLogger: params.disableLogger || pkg.rooseveltConfig.disableLogger || false,
     noMinify: params.noMinify || pkg.rooseveltConfig.noMinify || false,
     enableValidator: params.enableValidator || pkg.rooseveltConfig.enableValidator || false,
-    htmlValidator: params.htmlValidator || pkg.rooseveltConfig.htmlValidator || {port: '8888', format: 'text', suppressWarnings: false},
+    htmlValidator: params.htmlValidator || pkg.rooseveltConfig.htmlValidator || {port: '8888', validatorAlwaysOn: false, format: 'text', suppressWarnings: false},
```
Comment (kethinov) [inline lib/sourceParams.js]: - change `validatorAlwaysOn` to `separateProcess`
- document this in the README
- add an ability to enable/disable this param via command line as well

Comment (kethinov) [inline lib/sourceParams.js]: - in addition to command line argument also support enable/disable via an environment variable

File: package.json [hunk 2]
```diff
@@ -139,6 +139,7 @@
   "gitHead": "",
   "scripts": {
     "test": "./node_modules/.bin/eslint .",
+    "killValidator": "node ./lib/killValidator.js",
```
Comment (alallier) [inline package.json]: I think you mean to put this in the roosevelt generator

File: README.md [hunk 3]
```diff
@@ -133,6 +133,24 @@ Run your app in production mode:
 npm run prod
 ```
 
+Run your app with an attached HTML Validator (HTML Validator is attached by default):
```
Comment (kethinov) [inline README.md]: decapitalize validator all over

File: README.md [hunk 4]
```diff
@@ -133,6 +133,24 @@ Run your app in production mode:
...
+
+```
+npm run dev detach
+```
+
+After running your app with a detached HTML Validator, use this command to shut down the HTML Validator:
+
+```
+npm run killValidator
+```
```
Comment (kethinov) [inline README.md]: Add this here:

*Note: if your Roosevelt app was created with an older version, you may not have this command in your npm scripts. If so, then manually add this npm script to your package.json: `"killValidator": "node ./node_modules/roosevelt/lib/killValidator.js"`*

File: README.md [hunk 5]
```diff
@@ -133,6 +133,24 @@ Run your app in production mode:
 npm run prod
 ```
 
+Run your app with an attached HTML Validator (HTML Validator is attached by default):
+
+```
+npm run dev attach
+```
+
+Run your app with a detached HTML Validator:
```
Comment (kethinov) [inline README.md]: Run your app with a detached HTML validator (runs validator as a separate process in the background; process must be manually killed later, see below):

File: README.md [hunk 6]
```diff
@@ -223,9 +241,11 @@ App behavior parameters
   - Use the environment variable `ROOSEVELT_ENABLE_VALIDATOR` to force the validator on or off regardless of dev or prod mode
   - e.g. force validator off in dev mode regardless of app settings:  `export ROOSEVELT_ENABLE_VALIDATOR=false && npm run dev`
   - e.g. force validator on in prod mode regardless of app settings:  `export ROOSEVELT_ENABLE_VALIDATOR=true && npm run prod`
-- `htmlValidator`: Params to send to [html-validator](https://github.com/zrrrzzt/html-validator#usage) (if `enableValidator` is set to true). When `suppressWarnings` is set to true validation warnings will be hidden and only errors will be shown.
-  - Default:  `{port: '8888', format: 'text', suppressWarnings: false}`
+- `htmlValidator`: Params to send to [html-validator](https://github.com/zrrrzzt/html-validator#usage) (if `enableValidator` is set to true). When `suppressWarnings` is set to true validation warnings will be hidden and only errors will be shown. When `separateProcess` is turned to true the HTML validator will run separately from the node process.
```
Comment (kethinov) [inline README.md]: change "is turned to true" to "is set to true"

File: README.md [hunk 7]
```diff
@@ -223,9 +241,11 @@ App behavior parameters
...
+  - Default:  `{port: '8888', separateProcess: false, format: 'text', suppressWarnings: false}`
   - Can be disabled for individual requests by sending the request header `Partial` with the value set to `true` or by passing `_disableValidator` to the model and setting it to `true`.
+  - e.g. force validator to detach regardless of app settings: `export ENABLE_DETACH=true && npm run dev`
```
Comment (kethinov) [inline README.md]: follow the convention above with another bullet explaining the env var in general:

- Use the environment variable `ROOSEVELT_ENABLE_DETACH` to force the validator to be an attached or detached process

File: lib/htmlValidator.js [hunk 8]
```diff
@@ -22,12 +23,32 @@ module.exports = function(app, callback) {
...
       htmlArrLength,
+      isError = false,
+      calledKill = false,
       htmlLine,
       formattedHTML,
       validatorTimeout,
       errorPage = path.normalize(__dirname + '/../defaultErrorPages/views/validatorErrors'),
       validatorDisabled;
 
+  if (process.env.ENABLE_DETACH === 'true') {
```
Comment (kethinov) [inline lib/htmlValidator.js]: make it `ROOSEVELT_ENABLE_DETACH`

File: lib/killValidator.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    const { statusCode } = res;
+
+    let error,
+        rawData = '';
+    if (statusCode !== 200) {
+      error = new Error('Request Failed.\n' +
+                        `Status Code: ${statusCode}`);
+    }
+    if (error) {
+      // consume 404 respone data
```
Comment (Autre31415) [inline lib/killValidator.js]: type here: 'response'

File: lib/killValidator.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    res.setEncoding('utf8');
+
+    res.on('data', (chunk) => {
+      rawData += chunk;
+    });
+
+    res.on('end', () => {
+      try {
+        // console.log(rawData);
```
Comment (Autre31415) [inline lib/killValidator.js]: old console.log comment

File: lib/killValidator.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            rawData += chunk;
+          });
+
+          res.on('end', () => {
+            try {
+              if (rawData.includes('<title>Ready to check  - Nu Html Checker</title>')) {
+                console.log('✔️  ' + ('Validator succesfully found on port: ' + currentPort).green);
+                foundValidator = true;
+                killValidator(currentPort);
+                // process.exit();
```
Comment (Autre31415) [inline lib/killValidator.js]: dev comment here too

**Pull request merged**